### ISSUE TITLE
M4/2.10 quasiquotes compiler error

### ISSUE THREAD

**Issue by etorreborre** [scalamacros/paradise#24]
Following [this suggestion](https://groups.google.com/d/msg/scala-language/0PhHUF8bFDQ/Y_rmhMi-PTIJ) I tried to move my code to use the `StandardLiftables` but I end up with a compiler error:

``` sh
error]
[error]      while compiling: /Users/etorreborre/projects/specs2/specs2/matcher-extra/src/main/scala/org/specs2/matcher/MatcherMacros.scala
[error]         during phase: typer
[error]      library version: version 2.10.3
[error]     compiler version: version 2.10.3
[error]   reconstructed args: -Ywarn-nullary-unit -feature -Xlint -Ywarn-inaccessible -Xcheckinit -unchecked -classpath /Users/etorreborre/projects/specs2/specs2/matcher-extra/target/scala-2.10/classes:/Users/etorreborre/projects/specs2/specs2/analysis/target/scala-2.10/classes:/Users/etorreborre/projects/specs2/specs2/core/target/scala-2.10/classes:/Users/etorreborre/projects/specs2/specs2/matcher/target/scala-2.10/classes:/Users/etorreborre/projects/specs2/specs2/common/target/scala-2.10/classes:/Users/etorreborre/projects/specs2/specs2/scalacheck/target/scala-2.10/classes:/Users/etorreborre/.sbt/boot/scala-2.10.3/lib/scala-compiler.jar:/Users/etorreborre/.sbt/boot/scala-2.10.3/lib/scala-reflect.jar:/Users/etorreborre/.ivy2/cache/org.scalaz/scalaz-core_2.10/bundles/scalaz-core_2.10-7.0.6.jar:/Users/etorreborre/.ivy2/cache/org.scalaz/scalaz-concurrent_2.10/bundles/scalaz-concurrent_2.10-7.0.6.jar:/Users/etorreborre/.ivy2/cache/org.scalaz/scalaz-effect_2.10/bundles/scalaz-effect_2.10-7.0.6.jar:/Users/etorreborre/.ivy2/cache/org.specs2/classycle/jars/classycle-1.4.3.jar:/Users/etorreborre/.ivy2/cache/org.scalacheck/scalacheck_2.10/jars/scalacheck_2.10-1.11.3.jar:/Users/etorreborre/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar:/Users/etorreborre/.ivy2/cache/org.scalamacros/quasiquotes_2.10.3/jars/quasiquotes_2.10.3-2.0.0-M5.jar -Xplugin:/Users/etorreborre/.ivy2/cache/org.scalamacros/paradise_2.10.3/jars/paradise_2.10.3-2.0.0-M5.jar -deprecation -bootclasspath /Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/sunrsasign.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.7.0_51.jdk/Contents/Home/jre/classes:/Users/etorreborre/.sbt/boot/scala-2.10.3/lib/scala-library.jar -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -language:_ -Ywarn-nullary-override -Ywarn-adapted-args
[error]
[error]   last tree to typer: Apply(method apply)
[error]               symbol: method apply in trait SyntacticClassDefExtractor (flags: <method> <deferred> <triedcooking>)
[error]    symbol definition: def apply(mods: QuasiquoteCompat.this.u.Modifiers,name: QuasiquoteCompat.this.u.TypeName,tparams: List[QuasiquoteCompat.this.u.Tree],constrMods: QuasiquoteCompat.this.u.Modifiers,vparamss: List[List[QuasiquoteCompat.this.u.Tree]],earlyDefs: List[QuasiquoteCompat.this.u.Tree],parents: List[QuasiquoteCompat.this.u.Tree],selfType: QuasiquoteCompat.this.u.Tree,body: List[QuasiquoteCompat.this.u.Tree]): QuasiquoteCompat.this.u.ClassDef
[error]                  tpe: _1829.u.ClassDef
[error]        symbol owners: method apply -> trait SyntacticClassDefExtractor -> trait ReificationSupportApi -> trait QuasiquoteCompat -> package quasiquotes
[error]       context owners: value matchers -> method matcherMacroImpl -> object MatcherMacros -> package matcher
[error]
[error] == Enclosing template or block ==
[error]
[error] Import( // val <import>: ImportType(c.universe)
[error]   "c"."universe" // val universe: scala.reflect.macros.Universe in trait Context, tree.tpe=c.universe.type
[error]   List(
[error]     ImportSelector(_,978,null,-1)
[error]   )
[error] )
[error]
[error] == Expanded type of tree ==
[error]
[error] AbstractTypeRef(
[error]   pre = SingleType(
[error]     pre = AbstractTypeRef(
[error]       AbstractType(
[error]         type _1829.type <: scala.quasiquotes.QuasiquoteCompat{val u: MakeMatchers.this.c.universe.type} with Singleton
[error]
[error]       )
[error]     )
[error]     value u
[error]   )
[error]   AbstractType(
[error]     type ClassDef >: Null <: Trees.this.ImplDef with Trees.this.ClassDefApi
[error]
[error]   )
[error] )
[error]
[error] uncaught exception during compilation: java.lang.AssertionError
[trace] Stack trace suppressed: run last specs2-matcher-extra/compile:compile for the full output.
[error] (specs2-matcher-extra/compile:compile) java.lang.AssertionError: assertion failed: mkAttributedQualifier(_3003.type, <none>)
```

**Comment by stanch** [comment]
Could you post the code? Maybe I could help, having insomnia anyway :)

**Comment by stanch** [comment]
https://github.com/etorreborre/specs2/blob/b767765a3512c4fd9adbe4007bcf1f0091a68984/matcher-extra/src/main/scala/org/specs2/matcher/MatcherMacros.scala#L30

Should not this be

``` scala
new MakeMatchers(c).matchers[T]
                            ^^^
```

(Just a random thought)

**Comment by etorreborre** [comment]
You're right, the annotation is missing. But then the error becomes:

``` sh
error]   last tree to typer: Ident(T)
[error]               symbol: type T (flags: <deferred> <param>)
[error]    symbol definition: T
[error]                  tpe: T
[error]        symbol owners: type T -> method matcherMacroImpl -> object MatcherMacros -> package matcher
[error]       context owners: value matchers -> method matcherMacroImpl -> object MatcherMacros -> package matcher
[error]
[error] == Enclosing template or block ==
[error]
[error] Import( // val <import>: ImportType(c.universe)
[error]   "c"."universe" // val universe: scala.reflect.macros.Universe in trait Context, tree.tpe=c.universe.type
[error]   List(
[error]     ImportSelector(_,978,null,-1)
[error]   )
[error] )
[error]
[error] == Expanded type of tree ==
[error]
[error] SkolemTypeRef(TypeSkolem(T))
[error]
[error] uncaught exception during compilation: java.lang.AssertionError
[trace] Stack trace suppressed: run last specs2-matcher-extra/compile:compile for the full output.
[error] (specs2-matcher-extra/compile:compile) java.lang.AssertionError: assertion failed: mkAttributedQualifier(_3003.type, <none>)
```

**Comment by stanch** [comment]
https://github.com/etorreborre/specs2/blob/b767765a3512c4fd9adbe4007bcf1f0091a68984/matcher-extra/src/main/scala/org/specs2/matcher/MatcherMacros.scala#L34

Ok, next question: where does `matcherClassType` come from? (I know it’s in `MakeMatchers`)
Kinda strange though, that it compiles (oh, wait, it does not...)

**Comment by etorreborre** [comment]
Doh! Ok, fixed in [de91850](https://github.com/etorreborre/specs2/commit/de918508d52caa1062e40aafaee95ddcc9a87783). With the same error though. But maybe my "fix" is not correct anyway. I'm actually doing other things at the moment and should take some time to rewrite all of this properly with `StandardLiftables`.

**Comment by xeno-by** [comment]
Thanks for reporting guys! Normally I'd get right at fixing this stuff, but I'm completely tied up in the beginning of this week, so it'll have to wait a bit. I'll do my best to take a short look today or tomorrow though.

**Comment by etorreborre** [comment]
No worries, take your time.

**Comment by xeno-by** [comment]
Okay, I know what the problem is, but to solve it I need to have the compiler plugin depend on the quasiquote library. Unfortunately, sbt won't understand this sort of dependencies, so I need to merge `paradise.jar` and `quasiquotes.jar` into `paradise.jar` and publish the merged jar.

Here's my naive stab at it: https://github.com/scalamacros/paradise/blob/95cfba025b452d5bc0b65004cab307ffee41118f/project/build.scala#L129, however it doesn't work - after `package` I have the correct jar, but `publish` publishes a slim jar. Any ideas on how to implement this?

Unfortunately I have already overextended my time budget for anything not EPFL-related for today and tomorrow, so I'd really appreciate your help here.

**Comment by xeno-by** [comment]
@etorreborre Okay, I fixed the blocker that Nick has reported in the nearby thread. Hopefully will be able take a look at this one soon. Am I right in understanding that it only crashes on the code that wouldn't have compiled anyway? Also, what sha do I need to check out to reproduce this?

**Comment by etorreborre** [comment]
@xeno-by the latest branch I have, which I think is compiling and crashing is [de9185](https://github.com/etorreborre/specs2/commit/de918508d52caa1062e40aafaee95ddcc9a87783).

**Comment by xeno-by** [comment]
@etorreborre Explicitly specifying the type argument for `MakeMatcher` as outlined below fixes the problem. Looks like a bug in Scala's implicit search. I'll be closing this issue and will try to compose a minimized reproduction for JIRA.

```
val (matcherClassType, classDefinition) = new MakeMatchers[c.type](c).matchers[T]
```

**Comment by etorreborre** [comment]
Thanks Eugene for the debugging, this works fine now (and thanks @stanch for having a look at it too)

**Comment by xeno-by** [comment]
Thank you very much for your help!

**Comment by etorreborre** [comment]
Actually I have one more question on M5. In 2.10.3 `StandardLiftables` comes from `org.scala.quasiquotes`. But there is no `org.scala.quasiquotes` in 2.11 so my code is not compiling anymore for 2.11. What should I do?

**Comment by xeno-by** [comment]
In M4 and M5 it's `scala.quasiquotes`. I had to change the package name to access some important `private[scala]` methods.

**Comment by etorreborre** [comment]
I mean what is it in 2.11? How can I keep the same code under both 2.10 and 2.11?

**Comment by etorreborre** [comment]
At the moment the build is [failing](https://travis-ci.org/etorreborre/specs2/jobs/21546675) with 2.11 because the `org.scala.quasiquotes` package is unknown.

**Comment by etorreborre** [comment]
Should we reopen this issue or open another one? Because unless I'm wrong this is going to be a problem for codebases across 2.10 and 2.11.

**Comment by xeno-by** [comment]
I guess this falls under https://github.com/scalamacros/paradise/issues/20

**Comment by xeno-by** [comment]
Yes, indeed, this is not source compatible with 2.11. In 2.11, StandardLiftables are part of the universe cake, and programmers need not to refer to them - they are enabled by default. I'll see what can be done about that, so that we get the same beavior in 2.10.

**Comment by etorreborre** [comment]
Cool, thank you.

**Comment by gkossakowski** [comment]
The source incompatiblity issue made community build fail due to specs2. We've stopped tracking specs2 master to remedy the situation: https://github.com/typesafehub/community-builds/pull/29

**Comment by xeno-by** [comment]
I planned to fix source incompatibilities in 2.0.0-M6, which I hope to look into this weekend.
